core/src/main/scala/com/eed3si9n/jarjarabrams/Utils.scala

```diff
@@ -1,9 +0,0 @@
-package com.eed3si9n.jarjarabrams
-
-import java.io.InputStream
-
-object Utils {
-  // Use InputStream.readAllBytes() once on JDK >=9
-  def readAllBytes(inputStream: InputStream): Array[Byte] =
-    Stream.continually(inputStream.read).takeWhile(_ != -1).map(_.toByte).toArray
```

Comment:
This was reading things byte by byte, boxing each byte along the way…

Reply:
Ok input streams are now give from here, but I'm going to have to achieve the same (read all the bytes) in Mill, how can I achieve this better without Java 9?

Also did you see my comment https://github.com/lihaoyi/mill/pull/947#discussion_r473646741, before in Mill we were not loading all the jars at once in memory thanks to the `InputStream`. It sounds like we have no choice when shading but I'd like at least avoid loading when no shading rules are given. Do you recommend reading all the bytes, shading and going back to a `ByteArrayInputStream` in Mill's codebase?

Reply:
> how can I achieve this better without Java 9?

I usually do it [this way](https://github.com/coursier/coursier/blob/f8c2f7a20de86b90719e281f04f0fadf53fc8070/modules/tests/jvm/src/test/scala/coursier/Platform.scala#L13-L25). I never really benchmarked that, but this doesn't box and wrap in a `Stream` every `Byte` of the `InputStream`.

Reply:
> Do you recommend only when shading rules is not empty, reading all the bytes, shading and going back to a ByteArrayInputStream in Mill's codebase?

Isn't it possible to keep things as `(String, () => InputStream)` in mill? If the rules are empty, we could keep the same `() => InputStream` as before (something like `() => zipFile.getInputStream(entry)` IIRC). If the rules are non-empty, then we could keep things in memory in a first time, with a `() => new ByteArrayInputStream(bytes)` (with all the `bytes` living in memory…).

I had missed that in the mill code base, all the names / mappings are needed in an eager collection, before "grouping", which prevents keeping things in `Iterator`s / `Generator`s.

Ideally, the shading library here should allow us to independently get the new name and process the byte code… That way, the byte code processing could be delayed in the `() => InputStream`. I guess this can be done later…

Reply:
> Isn't it possible to keep things as (String, () => InputStream) in mill?

What is the point of `() => InputStream` as opposed to `InputStream`? An input stream is not taking up memory but already describes how to access data as a stream.

> I had missed that in the mill code base, all the names / mappings are needed in an eager collection, before "grouping", which prevents keeping things in Iterators / Generators.

Yes that's the annoying part here since we need the ability to "append" files.

> Ideally, the shading library here should allow us to independently get the new name and process the byte code… That way, the byte code processing could be delayed in the `() => InputStream`. I guess this can be done later…

I was looking at this too but there is not immediate easy way to do this.